shelley-ma/shelley-ma-test/src/Test/Cardano/Ledger/ShelleyMA/Serialisation/Generators.hs

```diff
@@ -1,51 +1,334 @@
+import Cardano.Binary
+  ( toCBOR,
+  )
+import Cardano.Crypto.Hash (HashAlgorithm, hashWithSerialiser)
+import qualified Cardano.Crypto.Hash as Hash
+import Cardano.Ledger.Allegra (AllegraEra)
+import qualified Cardano.Ledger.Core as Abstract
+import Cardano.Ledger.Era (Era)
+import Cardano.Ledger.Mary (MaryEra)
+import qualified Cardano.Ledger.Mary.Value as Mary (AssetID (..), PolicyID (..), Value (..))
+import Cardano.Ledger.Shelley (ShelleyBased)
+import qualified Cardano.Ledger.ShelleyMA.Rules.Utxo as MA.STS
+import qualified Cardano.Ledger.ShelleyMA.Scripts as MA (Timelock (..))
+import Cardano.Ledger.ShelleyMA.Timelocks (Timelock (..), ValidityInterval)
+import qualified Cardano.Ledger.ShelleyMA.TxBody as MA (TxBody (..))
+import Data.ByteString.Char8 (ByteString)
+import Data.Coerce (coerce)
+import Data.Sequence.Strict (StrictSeq, fromList)
+import Generic.Random (genericArbitraryU)
+import Shelley.Spec.Ledger.API hiding (SignedDSIGN, TxBody (..))
+import Shelley.Spec.Ledger.Coin (DeltaCoin)
+import Shelley.Spec.Ledger.MetaData (MetaDataHash)
+import qualified Shelley.Spec.Ledger.STS.Ledger as STS
+import qualified Shelley.Spec.Ledger.STS.Ledgers as STS
+import qualified Shelley.Spec.Ledger.STS.Ppup as STS
+import qualified Shelley.Spec.Ledger.STS.Utxow as STS
+import Shelley.Spec.Ledger.Slot (SlotNo)
+import Test.QuickCheck
+  ( Arbitrary,
+    arbitrary,
+    choose,
+    genericShrink,
+    listOf,
+    oneof,
+    resize,
+    shrink,
   )
-import Data.Sequence.Strict (fromList)
-import Shelley.Spec.Ledger.Keys (KeyHash (..))
-import Shelley.Spec.Ledger.BaseTypes (StrictMaybe (SJust, SNothing))
-import Test.Shelley.Spec.Ledger.Serialisation.Generators(mkDummyHash) -- imports arbitray instance for MultiSig
-import Test.Tasty.QuickCheck hiding (scale)
+import Test.Shelley.Spec.Ledger.ConcreteCryptoTypes (Mock)
+import Test.Tasty.QuickCheck (Gen)
 
--- ================================================================
+{-------------------------------------------------------------------------------
+  ShelleyMAEra Generators
 
--- ================================================================================
--- Some generators for Timelock
+  Generators used for roundtrip tests, generated values are not
+  necessarily valid
+-------------------------------------------------------------------------------}
 
-genSlot :: Gen (StrictMaybe SlotNo)
-genSlot = oneof [ pure SNothing, (SJust . SlotNo) <$> choose (0,10)]
+mkDummyHash :: forall h a. HashAlgorithm h => Int -> Hash.Hash h a
+mkDummyHash = coerce . hashWithSerialiser @h toCBOR
+
+genTxBody ::
+  ( ShelleyBased era,
+    Arbitrary (Abstract.Value era),
+    Arbitrary (TxIn era),
+    Arbitrary (StrictSeq (TxOut era)),
+    Arbitrary (StrictSeq (DCert era)),
+    Arbitrary (Wdrl era),
+    Arbitrary Coin,
+    Arbitrary (StrictMaybe (Update era)),
+    Arbitrary (StrictMaybe (MetaDataHash era)),
+    Arbitrary (StrictMaybe SlotNo)
```

Comment:
I think you can simplify these constraints, e.g., `Arbitrary (StrictSeq (TxOut era))` -> `Arbitrary (TxOut era)` and `Arbitrary (StrictMaybe a)` -> `Arbitrary a`. Enabling the [`Wredundant-constraints`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-warnings.html?highlight=redundant%20constraints#ghc-flag--Wredundant-constraints) or [`-Wsimplifiable-class-constraints`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-warnings.html?highlight=simplifiable#ghc-flag--Wsimplifiable-class-constraints) constraints should warn you about this. The same is true for a bunch of other instances.

Reply:
also here, the `StrictSeq` is defined in EraIndep which I don't ref here, which prevents the simplifying to `Arbitrary (TxOut era)`